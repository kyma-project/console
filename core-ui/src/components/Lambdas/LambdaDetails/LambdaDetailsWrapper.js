import React from 'react';
import LuigiClient from '@kyma-project/luigi-client';
import PropTypes from 'prop-types';

import { useQuery } from '@apollo/react-hooks';
import { GET_LAMBDA } from '../../../gql/queries';
import { REFETCH_TIMEOUT } from '../../../shared/constants';

import EntryNotFound from '../../EntryNotFound/EntryNotFound';
import { Spinner } from 'react-shared';
import LambdaDetails from './LambdaDetails';

LambdaDetailsWrapper.propTypes = {
  lambdaName: PropTypes.string.isRequired,
};

export default function LambdaDetailsWrapper({ lambdaName }) {
  const namespace = LuigiClient.getEventData().environmentId;
  const { data, error, loading, refetch } = useQuery(GET_LAMBDA, {
    variables: {
      name: lambdaName,
      namespace,
    },
    fetchPolicy: 'no-cache',
  });

  if (error) {
    return `Error! ${error.message}`;
  }
  if (loading) {
    return <Spinner />;
  }
  if (data && !data.function) {
    setTimeout(() => {
      refetch();
    }, REFETCH_TIMEOUT);
    return <EntryNotFound entryType="Lambda" entryId={lambdaName} />;
  }
  if (data && data.function) {
    return <LambdaDetails lambda={data.function} />;
  }
}
