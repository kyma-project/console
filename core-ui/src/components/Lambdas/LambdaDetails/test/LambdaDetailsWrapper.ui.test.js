import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';

import { lambda } from './gqlMocks';
import { componentUpdate } from '../../../../testing';
import LambdaDetailsWrapper from '../LambdaDetailsWrapper';
import { Spinner } from 'react-shared';

jest.mock('@kyma-project/luigi-client', () => {
  return {
    getEventData: () => ({ environmentId: 'testnamespace' }),
    getNodeParams: () => ({ selectedTab: 'Code' }),
  };
});

describe('LambdaDetailsWrapper', () => {
  it('Renders with an error in no lambdaName is provided', async () => {
    console.error = jest.fn();
    let component;
    await renderer.act(async () => {
      component = renderer.create(
        <MockedProvider>
          <LambdaDetailsWrapper />
        </MockedProvider>,
      );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(console.error.mock.calls.length).toBe(1);
    expect(console.error.mock.calls[0][0]).toMatchSnapshot();
  });

  it('Shows loading indicator only when data is not yet loaded', async () => {
    const component = mount(
      <MockedProvider>
        <LambdaDetailsWrapper lambdaName={lambda.name} />
      </MockedProvider>,
    );
    expect(component.find(Spinner)).toHaveLength(1);

    await componentUpdate(component);

    expect(component.find(Spinner)).toHaveLength(0);
  });
});
