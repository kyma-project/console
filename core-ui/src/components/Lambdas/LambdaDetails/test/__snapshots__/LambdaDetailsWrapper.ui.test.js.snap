// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaDetailsWrapper Renders with an error in no lambdaName is provided 1`] = `
"Error! Network error: No more mocked responses for the query: query Function($name: String!, $namespace: String!) {
  function(name: $name, namespace: $namespace) {
    name
    namespace
    labels
    runtime
    size
    status
    content
    dependencies
    serviceBindingUsages {
      name
      parameters {
        envPrefix {
          name
          __typename
        }
        __typename
      }
      serviceBinding {
        name
        serviceInstanceName
        secret {
          name
          data
          __typename
        }
        __typename
      }
      __typename
    }
    __typename
  }
}
, variables: {\\"namespace\\":\\"testnamespace\\"}"
`;

exports[`LambdaDetailsWrapper Renders with an error in no lambdaName is provided 2`] = `
"Warning: Failed prop type: The prop \`lambdaName\` is marked as required in \`LambdaDetailsWrapper\`, but its value is \`undefined\`.
    in LambdaDetailsWrapper (at LambdaDetailsWrapper.ui.test.js:25)"
`;
